src/markdown/generator.ml

```diff
@@ -0,0 +1,382 @@
+
+  let anchor' s = Anchor s
+
+  let string s = String s
+
+  let block' ts = Block ts
+
+  let inline' l = List.map (fun s -> string s) l |> concat
+
+  let str fmt = Format.ksprintf (fun s -> string s) fmt
+
+  let escaped fmt = Format.ksprintf (fun s -> string s) fmt
+
+  let rec pp fmt t =
+    match t with
+    | Block b ->
+        let inner = function
+          | [] -> ()
+          | [ x ] -> Format.fprintf fmt "%a" pp x
+          | x :: xs -> Format.fprintf fmt "%a@\n%a" pp x pp (Block xs)
+        in
+        inner b
+    | Concat l -> pp_many fmt l
+    | Break -> Format.fprintf fmt "@\n"
+    | Space -> Format.fprintf fmt " "
+    | Anchor s -> Format.fprintf fmt "<a id=\"%s\"></a>" s
+    | String s -> Format.fprintf fmt "%s" s
+    (* We use double backticks to take care of polymorphic variants or content
+       within backtick, and the spaces before and after the backticks for
+       clarity on what should be enclosed in backticks. For example,
+       "type nums = [ | `One | `Two ]" would be rendered as "``|`````Monday`` "
+       if the spaces were missing.
+    *)
+    | Backticks -> Format.fprintf fmt " `` "
+    | Nbsp -> Format.fprintf fmt "\u{00A0}"
+    | OpenSqBracket -> Format.fprintf fmt "["
+    | CloseSqBracket -> Format.fprintf fmt "]"
+    | OpenParenthesis -> Format.fprintf fmt "("
+    | CloseParenthesis -> Format.fprintf fmt ")"
+
+  and pp_many fmt l = List.iter (pp fmt) l
+end
+
+open Markup
+
+let entity e =
+  match e with "#45" -> escaped "-" | "gt" -> str ">" | s -> str "&%s;" s
+
+let raw_markup (_ : Raw_markup.t) = noop
+
+let style (style : style) content =
+  match style with
+  | `Bold -> string "**" ++ (content ++ str "**")
+  | `Italic | `Emphasis -> string "_" ++ (content ++ str "_")
+  | `Superscript -> string "<sup>" ++ content
+  | `Subscript -> string "<sub>" ++ content
+
+let make_hashes n = String.make n '#'
+
+type args = { generate_links : bool ref; md_flavour : string ref }
+
+let args = { generate_links = ref true; md_flavour = ref "" }
+
+let rec source_code (s : Source.t) nbsp =
+  match s with
+  | [] -> noop
+  | h :: t -> (
+      let continue s = if s = [] then concat [] else source_code s nbsp in
+      match h with
+      | Source.Elt i -> inline i nbsp ++ continue t
+      | Tag (None, s) -> continue s ++ continue t
+      | Tag (Some _, s) -> continue s ++ continue t)
+
+and inline (l : Inline.t) nbsp =
+  match l with
+  | [] -> noop
+  | i :: rest -> (
+      let continue i = if i = [] then noop else inline i nbsp in
+      let make_link c s =
+        open_sq_bracket ++ continue c ++ close_sq_bracket ++ open_parenthesis
+        ++ string s ++ close_parenthesis ++ continue rest
+      in
+      let cond then_clause else_clause =
+        if !(args.generate_links) then then_clause else else_clause
+      in
+      match i.desc with
+      | Text "" -> continue rest
+      | Text s -> (
+          match s with
+          | "end" | "}" | "]" ->
```

Comment:
Documentation containing these words is broken:
```
(** end } ] *)
```

The keywords must not be matched into any text, only when nested into a `source_code` with the right tag.

Reply:
Hmm! only `}` behaves not right but yes, that’s a good catch. Thanks!